outlines/models/exllamav2.py

```diff
@@ -1,121 +1,217 @@
-                    preprocess_only=True,
-                    loras=[self.lora],
+        self.generator = generator
+        self.tokenizer = tokenizer
+        self.max_seq_len = max_seq_len
+
+    def prepare_generation_parameters(
+        self,
+        prompts: Union[str, List[str]],
+        generation_parameters: GenerationParameters,
+        sampling_parameters: SamplingParameters,
+        structure_logits_processor,
+        **exllamav2_params: Unpack[ExllamaV2Params],
+    ) -> Tuple[ExllamaV2Params, Union[str, List[str]]]:
+        """Prepare the generation parameters.
+
+        `exllamav2` uses different default values
+
+        """
+        from exllamav2.generator import ExLlamaV2Sampler
+
+        if isinstance(prompts, str):
+            prompts = [prompts]
+        max_tokens, stop_at, seed = dataclasses.astuple(generation_parameters)
+
+        if max_tokens is None:
+            max_tokens = []
+            for prompt in prompts:
+                ids = self.generator.tokenizer.encode(
+                    prompt, encode_special_tokens=False
                 )
+                prompt_tokens = ids.shape[-1]
+                max_tokens.append(self.max_seq_len - prompt_tokens)
+            exllamav2_params["max_new_tokens"] = max_tokens
+        else:
+            exllamav2_params["max_new_tokens"] = [
+                max_tokens for _ in range(len(prompts))
+            ]
+
+        stop_conditions = [self.generator.tokenizer.eos_token_id]
+        if isinstance(generation_parameters.stop_at, str):
+            stop_conditions.append(generation_parameters.stop_at)
+        elif isinstance(generation_parameters.stop_at, list):
+            for stop_at in generation_parameters.stop_at:
+                stop_conditions.append(stop_at)
+        exllamav2_params["stop_conditions"] = stop_conditions
+        exllamav2_params["seed"] = seed
+
+        gen_settings = ExLlamaV2Sampler.Settings()
+        if sampling_parameters.temperature is not None:
+            gen_settings.temperature = sampling_parameters.temperature
+        if sampling_parameters.top_p is not None:
+            gen_settings.top_p = sampling_parameters.top_p
+        if sampling_parameters.top_k is not None:
+            gen_settings.top_k = sampling_parameters.top_k
+        gen_settings.logits_processor = structure_logits_processor
+        exllamav2_params["gen_settings"] = gen_settings
+        if sampling_parameters.num_samples > 1:
+            prompts = prompts * sampling_parameters.num_samples
+            exllamav2_params["max_new_tokens"] = (
+                exllamav2_params["max_new_tokens"] * sampling_parameters.num_samples
+            )
+
+        if len(prompts) == 1:
+            prompts = prompts[0]
 
-        self.past_seq = seq_tensor
+        return exllamav2_params, prompts
 
-        return self.model.forward(
-            seq_tensor[-1:].view(1, -1), self.cache, loras=[self.lora]
+    def reformat_output(
```

Comment:
why is `reformat_output` necessary? Could we get some docstrings and comments here and throughout. Especially for details which might not be understood by those making changes in the future?

Reply:
@lapp0 Ah so it's necessary mainly because for exllamav2, it mainly accepts only a list or a string then
1. If the prompt is a string, return a string
2. If a prompt is a list, return a list. This is not the same as outlines output in that if the list is only one element, the string is expected to be outputted.
4. No such thing as num_samples, so I had to duplicate the prompt by num_samples times and have this function output a list of lists
So in summary, it's to fix the output from exllamav2 to be in outlines/llamacpp form. I can add this as a comment if that works!

Reply:
@lapp0 Added comment! Let me know if it looks good